In [1]:
import glob
import pandas as pd
import numpy as np
import re
import nltk

Задание 1

Вернемся к набору данных о видеоиграх.

Ответьте на следующие вопросы:

1) Как критики относятся к спортивным играм?

2) Критикам нравятся больше игры на PC или на PS4?

3) Критикам больше нравятся стрелялки или стратегии?

Для каждого вопроса:

сформулируйте нулевую и альтернативную гипотезы;

выберите пороговый уровень статистической значимости;

опишите полученные результаты статистического теста.

#



Задание 1.1  Как критики относятся к спортивным играм?

In [2]:
vidgames = pd.read_csv('video_games_sales.csv')


In [3]:
vidgames.head(10)

Rank                                Name                      basename  \
0     1                          Wii Sports                    wii-sports   
1     2                   Super Mario Bros.              super-mario-bros   
2     3                      Mario Kart Wii                mario-kart-wii   
3     4       PlayerUnknown's Battlegrounds  playerunknowns-battlegrounds   
4     5                   Wii Sports Resort             wii-sports-resort   
5     6  Pokemon Red / Green / Blue Version                    pokmon-red   
6     7               New Super Mario Bros.          new-super-mario-bros   
7     8                              Tetris                        tetris   
8     9           New Super Mario Bros. Wii      new-super-mario-bros-wii   
9    10                           Minecraft                     minecraft   

          Genre ESRB_Rating Platform         Publisher              Developer  \
0        Sports           E      Wii          Nintendo           Nintendo EAD   
1      Platform         NaN      NES          Nintendo           Nintendo EAD   
2        Racing           E      Wii          Nintendo           Nintendo EAD   
3       Shooter         NaN       PC  PUBG Corporation       PUBG Corporation   
4        Sports           E      Wii          Nintendo           Nintendo EAD   
5  Role-Playing           E       GB          Nintendo             Game Freak   
6      Platform           E       DS          Nintendo           Nintendo EAD   
7        Puzzle           E       GB          Nintendo  Bullet Proof Software   
8      Platform           E      Wii          Nintendo           Nintendo EAD   
9          Misc         NaN       PC            Mojang              Mojang AB   

   VGChartz_Score  Critic_Score  ...  NA_Sales  PAL_Sales  JP_Sales  \
0             NaN           7.7  ...       NaN        NaN       NaN   
1             NaN          10.0  ...       NaN        NaN       NaN   
2             NaN           8.2  ...       NaN        NaN       NaN   
3             NaN           NaN  ...       NaN        NaN       NaN   
4             NaN           8.0  ...       NaN        NaN       NaN   
5             NaN           9.4  ...       NaN        NaN       NaN   
6             NaN           9.1  ...       NaN        NaN       NaN   
7             NaN           NaN  ...       NaN        NaN       NaN   
8             NaN           8.6  ...       NaN        NaN       NaN   
9             NaN          10.0  ...       NaN        NaN       NaN   

   Other_Sales    Year  Last_Update  \
0          NaN  2006.0          NaN   
1          NaN  1985.0          NaN   
2          NaN  2008.0  11th Apr 18   
3          NaN  2017.0  13th Nov 18   
4          NaN  2009.0          NaN   
5          NaN  1998.0          NaN   
6          NaN  2006.0          NaN   
7          NaN  1989.0          NaN   
8          NaN  2009.0          NaN   
9          NaN  2010.0  05th Aug 18   

                                                 url  status Vgchartzscore  \
0  http://www.vgchartz.com/game/2667/wii-sports/?...       1           NaN   
1  http://www.vgchartz.com/game/6455/super-mario-...       1           NaN   
2  http://www.vgchartz.com/game/6968/mario-kart-w...       1           8.7   
3  http://www.vgchartz.com/game/215988/playerunkn...       1           NaN   
4  http://www.vgchartz.com/game/24656/wii-sports-...       1           8.8   
5  http://www.vgchartz.com/game/4030/pokemon-red-...       1           NaN   
6  http://www.vgchartz.com/game/1582/new-super-ma...       1           NaN   
7  http://www.vgchartz.com/game/4534/tetris/?regi...       1           NaN   
8  http://www.vgchartz.com/game/35076/new-super-m...       1           9.1   
9  http://www.vgchartz.com/game/47724/minecraft/?...       1           NaN   

                                          img_url  
0   /games/boxart/full_2258645AmericaFrontccc.jpg  
1                    /games/boxart/8972270ccc.jpg  
2   /games/boxart/full_8932480Americ

Нулевая гипотеза - спортивным играм критики ставят в среднем 7.0 баллов и выше

Альтернативная гипотеза - спортивным играм критики ставят в среднем менее 7.0 баллов

alpha = 0.05

In [4]:
from scipy import stats as st
sport_gam = vidgames[vidgames.Genre=='Sports']

alpha = 0.05
print(sport_gam.Critic_Score.count())
print(sport_gam.Critic_Score.isnull().sum())


688
4556


В связи с тем, что количество оценок по спортивным играм от критиков составляет менее 15%, то заполнение 
пробелов каким-то значениями очень сильно исказит статистику. Поэтому пустые строки отбрасываем.

In [5]:
result = st.ttest_1samp(sport_gam.Critic_Score.dropna(), 7.0)
print(result)


Ttest_1sampResult(statistic=5.197156914436623, pvalue=2.671410902970997e-07)


Так как значение статистики больше нуля, то несмотря на маленькое значение p-value нулевую гипотезу не отбрасываем.

#


Задание 1.2 Критикам нравятся больше игры на PC или на PS4?

Нулевая гипотеза - критики относятся к PC играм так же как к играм на PS4

Альтернативная гипотеза - Критикам относятся по разному к играм на PC и PS4

alpha = 0.05

In [6]:
PC_gam = vidgames[vidgames.Platform	=='PC']
PS4_gam = vidgames[vidgames.Platform	=='PS4']

alpha = 0.05
print(PC_gam.Critic_Score.count())
print(PC_gam.Critic_Score.isnull().sum())
print(PS4_gam.Critic_Score.count())
print(PS4_gam.Critic_Score.isnull().sum())

865
10113
109
1646


В связи с тем, что количество оценок по обеим платформам от критиков составляет менее 10%, то заполнение пробелов каким-то значениями очень сильно исказит статистику. Поэтому пустые строки отбрасываем.

In [7]:
result = st.ttest_ind(PC_gam.Critic_Score.dropna(), PS4_gam.Critic_Score.dropna(), equal_var=False)
print(result)

Ttest_indResult(statistic=-2.7394476056351627, pvalue=0.006931808250254211)


Так как pvalue меньше уровня значимости, то нулевая гипотеза отвергается. Соответтсвенно критики относятся по разному к играм на PC и PS4 

# 


Задание 1.3 Критикам больше нравятся стрелялки или стратегии?

Нулевая гипотеза - критики относятся одинаково к стрелялкам и стратегиям

Альтернативная гипотеза - Критикам относятся по разному к стрелялкам и стратегиям

alpha = 0.05

In [8]:
Strat_gam = vidgames[vidgames.Genre	=='Strategy']
Shooter_gam = vidgames[vidgames.Genre	=='Shooter']

alpha = 0.05
print(Strat_gam.Critic_Score.count())
print(Strat_gam.Critic_Score.isnull().sum())
print(Shooter_gam.Critic_Score.count())
print(Shooter_gam.Critic_Score.isnull().sum())

369
2897
843
3743


В связи с тем, что количество оценок по обеим жанрам от критиков составляет менее 25%, то заполнение пробелов каким-то значениями очень сильно исказит статистику. Поэтому пустые строки отбрасываем.

In [9]:
result = st.ttest_ind(Strat_gam.Critic_Score.dropna(), Shooter_gam.Critic_Score.dropna(), equal_var=False)
print(result)

Ttest_indResult(statistic=1.6073949711166526, pvalue=0.10838786414223071)


Так как pvalue больше уровня значимости, то нулевая гипотеза не отбрасывается, и соответственно критики относятся и к стратегиям и к стрелялкам одинаково.

Задание 2

1) Привидите весь текст к нижнему регистру;

2) Удалите мусорные символы;

3) Удалите стоп-слова;

4) Привидите все слова к нормальной форме;

5) Преобразуйте все сообщения в вектора TF-IDF. Вам поможет следующий код:

from sklearn.feature_extraction.text import TfidfVectorizer

tfidf = TfidfVectorizer()
tfidf_matrix = tfidf.fit_transform(df.Message)
names = tfidf.get_feature_names()
tfidf_matrix = pd.DataFrame(tfidf_matrix.toarray(), columns=cname)
Можете поэкспериментировать с параметрами TfidfVectorizer;

6) Разделите данные на тестовые и тренировочные в соотношении 30/70, укажите random_state=42. Используйте train_test_split;

7) Постройте модель логистической регрессии, укажите random_state=42, оцените ее точность на тестовых данных;

8) Опишите результаты при помощи confusion_matrix;

9) Постройте датафрейм, который будет содержать все исходные тексты сообщений, классифицированные неправильно (с указанием фактического и предсказанного).

In [10]:
spam_df = pd.read_csv('spam.csv')

In [11]:
spam_df.head()

Category                                            Message
0      ham  Go until jurong point, crazy.. Available only ...
1      ham                      Ok lar... Joking wif u oni...
2     spam  Free entry in 2 a wkly comp to win FA Cup fina...
3      ham  U dun say so early hor... U c already then say...
4      ham  Nah I don't think he goes to usf, he lives aro...

In [12]:
# приводим сообщения к нижнему регистру и делим сразу на два дата фрейма, один данные, второй результат (к чему относятся данные)
messages = spam_df.Message.str.lower()
category = pd.DataFrame(spam_df.Category)

In [13]:
# удаляем мусорные символы
messages_wo_trash = messages.str.replace(r'[\W_]+',' ')

In [14]:
# разделяем на слова
splitted_messages = messages_wo_trash.str.split()
splitted_messages

0       [go, until, jurong, point, crazy, available, o...
1                          [ok, lar, joking, wif, u, oni]
2       [free, entry, in, 2, a, wkly, comp, to, win, f...
3       [u, dun, say, so, early, hor, u, c, already, t...
4       [nah, i, don, t, think, he, goes, to, usf, he,...
                              ...                        
5567    [this, is, the, 2nd, time, we, have, tried, 2,...
5568         [will, ü, b, going, to, esplanade, fr, home]
5569    [pity, was, in, mood, for, that, so, any, othe...
5570    [the, guy, did, some, bitching, but, i, acted,...
5571                     [rofl, its, true, to, its, name]
Name: Message, Length: 5572, dtype: object

In [15]:
from nltk.corpus import stopwords
stopwords_set = set(stopwords.words('english'))

In [16]:
def del_sw (spl_mes):
    sw_del = []
    for text in spl_mes:
        if text not in stopwords_set:
            sw_del.append(text)
    return sw_del

In [17]:
#удаляем стоп-слова
messages_without_sw = splitted_messages.apply(del_sw)
messages_without_sw

0       [go, jurong, point, crazy, available, bugis, n...
1                          [ok, lar, joking, wif, u, oni]
2       [free, entry, 2, wkly, comp, win, fa, cup, fin...
3           [u, dun, say, early, hor, u, c, already, say]
4          [nah, think, goes, usf, lives, around, though]
                              ...                        
5567    [2nd, time, tried, 2, contact, u, u, 750, poun...
5568                   [ü, b, going, esplanade, fr, home]
5569                            [pity, mood, suggestions]
5570    [guy, bitching, acted, like, interested, buyin...
5571                                   [rofl, true, name]
Name: Message, Length: 5572, dtype: object

In [18]:
from nltk.stem import WordNetLemmatizer
wordnet_lemmatizer = WordNetLemmatizer()

In [19]:
def lem_mes (mes_wo_sw):
    lem_words = []
    for word in mes_wo_sw:
        lem_words.append(wordnet_lemmatizer.lemmatize(word))
    return lem_words

In [20]:
lemmatized_message = messages_without_sw.apply(lem_mes)
lemmatized_message

0       [go, jurong, point, crazy, available, bugis, n...
1                          [ok, lar, joking, wif, u, oni]
2       [free, entry, 2, wkly, comp, win, fa, cup, fin...
3           [u, dun, say, early, hor, u, c, already, say]
4             [nah, think, go, usf, life, around, though]
                              ...                        
5567    [2nd, time, tried, 2, contact, u, u, 750, poun...
5568                   [ü, b, going, esplanade, fr, home]
5569                             [pity, mood, suggestion]
5570    [guy, bitching, acted, like, interested, buyin...
5571                                   [rofl, true, name]
Name: Message, Length: 5572, dtype: object

In [21]:
from sklearn.feature_extraction.text import TfidfVectorizer
#Данная функция не принимает списки из слов, а только строку
#tfidf_1 = TfidfVectorizer(lowercase=False)
#tfidf_matrix_1 = tfidf_1.fit_transform(lemmatized_message)
#cname = tfidf_1.get_feature_names()
#tfidf_matrix_1 = pd.DataFrame(tfidf_matrix_1.toarray(), columns=cname)

In [22]:
from gensim import corpora
# создадим словарь, в котором присвоем каждому словую свой уникальный id
dictionary = corpora.Dictionary(lemmatized_message)
dict(dictionary)

{0: 'amore',
 1: 'available',
 2: 'buffet',
 3: 'bugis',
 4: 'cine',
 5: 'crazy',
 6: 'e',
 7: 'go',
 8: 'got',
 9: 'great',
 10: 'jurong',
 11: 'la',
 12: 'n',
 13: 'point',
 14: 'wat',
 15: 'world',
 16: 'joking',
 17: 'lar',
 18: 'ok',
 19: 'oni',
 20: 'u',
 21: 'wif',
 22: '08452810075over18',
 23: '2',
 24: '2005',
 25: '21st',
 26: '87121',
 27: 'apply',
 28: 'c',
 29: 'comp',
 30: 'cup',
 31: 'entry',
 32: 'fa',
 33: 'final',
 34: 'free',
 35: 'may',
 36: 'question',
 37: 'rate',
 38: 'receive',
 39: 'std',
 40: 'text',
 41: 'tkts',
 42: 'txt',
 43: 'win',
 44: 'wkly',
 45: 'already',
 46: 'dun',
 47: 'early',
 48: 'hor',
 49: 'say',
 50: 'around',
 51: 'life',
 52: 'nah',
 53: 'think',
 54: 'though',
 55: 'usf',
 56: '1',
 57: '3',
 58: '50',
 59: 'back',
 60: 'chgs',
 61: 'darling',
 62: 'freemsg',
 63: 'fun',
 64: 'hey',
 65: 'like',
 66: 'rcv',
 67: 'send',
 68: 'still',
 69: 'tb',
 70: 'week',
 71: 'word',
 72: 'xxx',
 73: 'aid',
 74: 'brother',
 75: 'even',
 76: 'patent',


In [23]:
bows = [dictionary.doc2bow(text) for text in lemmatized_message]
bows

[[(0, 1),
  (1, 1),
  (2, 1),
  (3, 1),
  (4, 1),
  (5, 1),
  (6, 1),
  (7, 1),
  (8, 1),
  (9, 1),
  (10, 1),
  (11, 1),
  (12, 1),
  (13, 1),
  (14, 1),
  (15, 1)],
 [(16, 1), (17, 1), (18, 1), (19, 1), (20, 1), (21, 1)],
 [(22, 1),
  (23, 1),
  (24, 1),
  (25, 1),
  (26, 1),
  (27, 1),
  (28, 1),
  (29, 1),
  (30, 1),
  (31, 2),
  (32, 2),
  (33, 1),
  (34, 1),
  (35, 1),
  (36, 1),
  (37, 1),
  (38, 1),
  (39, 1),
  (40, 1),
  (41, 1),
  (42, 1),
  (43, 1),
  (44, 1)],
 [(20, 2), (28, 1), (45, 1), (46, 1), (47, 1), (48, 1), (49, 2)],
 [(7, 1), (50, 1), (51, 1), (52, 1), (53, 1), (54, 1), (55, 1)],
 [(18, 1),
  (39, 1),
  (56, 1),
  (57, 1),
  (58, 1),
  (59, 1),
  (60, 1),
  (61, 1),
  (62, 1),
  (63, 1),
  (64, 1),
  (65, 1),
  (66, 1),
  (67, 1),
  (68, 1),
  (69, 1),
  (70, 1),
  (71, 1),
  (72, 1)],
 [(65, 2), (73, 1), (74, 1), (75, 1), (76, 1), (77, 1), (78, 1)],
 [(79, 1),
  (80, 1),
  (81, 2),
  (82, 1),
  (83, 1),
  (84, 2),
  (85, 1),
  (86, 1),
  (87, 1),
  (88, 1),
  (89

In [24]:
from gensim.models import TfidfModel
tfidf = TfidfModel(bows)

# посмотрим, как это выглядит на "Происхождение видов"
tfidf[bows[5571]]

[(192, 0.4590231983055588),
 (1217, 0.5110627129481622),
 (3429, 0.7267135658919545)]

In [25]:
#не получается преобразовать модель gensim в датафрейм.
vector_df = pd.DataFrame(list(tfidf))

TypeError: 'int' object is not iterable

In [26]:
from sklearn.model_selection import train_test_split

In [27]:
#модель не хочет строиться на основании объекта gensim.models.tfidfmodel.TfidfModel, судя по всему надо использовать LDA именно из библиотеки gensim
X_train, X_test, y_train, y_test = train_test_split(tfidf, category, test_size=0.3, random_state=42)

TypeError: Expected sequence or array-like, got <class 'gensim.models.tfidfmodel.TfidfModel'>

In [28]:
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf = TfidfVectorizer(stop_words='english')
tfidf_matrix = tfidf.fit_transform(spam_df.Message)
cname = tfidf.get_feature_names()
tfidf_matrix = pd.DataFrame(tfidf_matrix.toarray(), columns=cname)

In [29]:
tfidf_matrix

00  000  000pes  008704050406  0089  0121  01223585236  01223585334  \
0     0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
1     0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
2     0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
3     0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
4     0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
...   ...  ...     ...           ...   ...   ...          ...          ...   
5567  0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
5568  0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
5569  0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
5570  0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   
5571  0.0  0.0     0.0           0.0   0.0   0.0          0.0          0.0   

      0125698789   02  ...  zhong  zindgi  zoe  zogtorius  zoom  zouk  zyada  \
0            0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   
1            0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   
2            0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   
3            0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   
4            0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   
...          ...  ...  ...    ...     ...  ...        ...   ...   ...    ...   
5567         0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   
5568         0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   
5569         0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   
5570         0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   
5571         0.0  0.0  ...    0.0     0.0  0.0        0.0   0.0   0.0    0.0   

       èn   ú1  〨ud  
0     0.0  0.0  0.0  
1     0.0  0.0  0.0  
2     0.0  0.0  0.0  
3     0.0  0.0  0.0  
4     0.0  0.0  0.0  
...   ...  ...  ...  
5567  0.0  0.0  0.0  
5568  0.0  0.0  0.0  
5569  0.0  0.0  0.0  
5570  0.0  0.0  0.0  
5571  0.0  0.0  0.0  

[5572 rows x 8440 columns]

In [30]:
X_train, X_test, y_train, y_test = train_test_split(tfidf_matrix, category.Category, test_size=0.3, random_state=42)

In [31]:
from sklearn.linear_model import LogisticRegression
lr = LogisticRegression(random_state=42)

In [32]:
lr.fit(X_train, y_train)

LogisticRegression(random_state=42)

In [33]:
lr.predict(X_test)

array(['ham', 'ham', 'ham', ..., 'ham', 'spam', 'ham'], dtype=object)

In [34]:
from sklearn.metrics import accuracy_score
accuracy_score(y_test, lr.predict(X_test))

0.9563397129186603

Аккуратность предсказания почти 96%

In [35]:
from sklearn.metrics import confusion_matrix
confusion_matrix(y_test, lr.predict(X_test))

array([[1447,    1],
       [  72,  152]], dtype=int64)

Из 1672 предсказаний: 1447 предсказаний ham было корректно, и 152 предсказния - spam.  Одно предсказание ham - было не верно и 72 предсказания spam было не верно, как видно из следующей таблицы для 72 результатов не была предсказана категория вовсе.

In [36]:
result = pd.DataFrame([y_test.values, lr.predict(X_test)]).T
result.head()

0    1
0  ham  ham
1  ham  ham
2  ham  ham
3  ham  ham
4  ham  ham

In [37]:
result_with_mes = pd.concat([result, spam_df.Message], join='inner', axis=1)

In [38]:
result_with_mes

0     1                                            Message
0      ham   ham  Go until jurong point, crazy.. Available only ...
1      ham   ham                      Ok lar... Joking wif u oni...
2      ham   ham  Free entry in 2 a wkly comp to win FA Cup fina...
3      ham   ham  U dun say so early hor... U c already then say...
4      ham   ham  Nah I don't think he goes to usf, he lives aro...
...    ...   ...                                                ...
1667   ham   ham  Guess who spent all last night phasing in and ...
1668  spam  spam  So now my dad is gonna call after he gets out ...
1669   ham   ham                        Yes..but they said its IT.,
1670  spam  spam  Very hurting n meaningful lines ever: "I compr...
1671   ham   ham                                        Lmao!nice 1

[1672 rows x 3 columns]

Судя по всему предсказать модель не смогла, там где используются не понятные междометия или вообще не английский язык.